### PR TITLE
Remove reserved condition

### PR DESCRIPTION
**Pull request by alenkacz**
Summary:
This diff removes all usages of `Condition.Reserved` and `Condition.ReservedTerminal`.
A migrate replaces these conditions with `Condition.Finished` on instance and task
states.

JIRA issues: MARATHON-8374

Co-authored-by: Karsten Jeschkies <k@jeschkies.xyz>

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdater.scala [hunk 1]
```diff
@@ -20,7 +20,7 @@ object InstanceUpdater extends StrictLogging {
 
     // We need to suspend reservation on already launched reserved instances
     // to prevent reservations being destroyed/unreserved.
-    val updatedReservation = if (updatedTask.status.condition == Condition.Reserved && !instance.reservation.exists(r => r.state.isInstanceOf[Reservation.State.Suspended])) {
+    val updatedReservation = if (updatedTask.status.condition.isTerminal && instance.hasReservation && !instance.reservation.exists(r => r.state.isInstanceOf[Reservation.State.Suspended])) {
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdater.scala]: Hm, how about `instance.isReserved`/`updatedTask.status.isReserved` with

```scala
def isReserved = isTerminal && hasReservation
```

It might not apply in this case but in others.

File: src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueActor.scala [hunk 2]
```diff
@@ -264,7 +264,7 @@ private[impl] class LaunchQueueActor(
 
       // Reuse resident instances that are stopped.
       val existingReservedStoppedInstances = await(instanceTracker.specInstances(runSpec.id))
-        .filter(i => i.isReserved && i.state.goal == Goal.Stopped) // resident to relaunch
+        .filter(i => i.hasReservation && i.state.goal == Goal.Stopped) // resident to relaunch
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueActor.scala]: I don't think this is correct. `isReserved` is defined as `isTerminal && hasReservation`. See my comment above.

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerDelegate.scala [hunk 3]
```diff
@@ -55,7 +55,7 @@ private[tracker] class InstanceTrackerDelegate(
   // TODO(jdef) support pods when counting launched instances
   override def countActiveSpecInstances(appId: PathId): Future[Int] = {
     import scala.concurrent.ExecutionContext.Implicits.global
-    instancesBySpec().map(_.specInstances(appId).count(instance => instance.isActive || (instance.isReserved && !instance.isReservedTerminal)))
+    instancesBySpec().map(_.specInstances(appId).count(instance => instance.isActive))
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerDelegate.scala]: This is so much simpler! 🎉

File: src/main/scala/mesosphere/marathon/core/task/update/impl/steps/ScaleAppUpdateStepImpl.scala [hunk 4]
```diff
@@ -22,7 +22,7 @@ class ScaleAppUpdateStepImpl @Inject() (
   private[this] lazy val schedulerActor = schedulerActorProvider.get()
 
   private[this] def scalingWorthy: Condition => Boolean = {
-    case Condition.Reserved | Condition.UnreachableInactive | _: Condition.Terminal => true
+    case Condition.UnreachableInactive | _: Condition.Terminal => true
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/task/update/impl/steps/ScaleAppUpdateStepImpl.scala]: Is this true? We have `ReservedTerminal` and `Reserved`. I'm not sure what `ReservedTerminal` was used for.

File: src/main/scala/mesosphere/marathon/raml/PodStatusConversion.scala [hunk 5]
```diff
@@ -249,36 +247,39 @@ trait PodStatusConversion {
     }.getOrElse(List.empty[ContainerEndpointStatus])
 
   def podInstanceState(
-    instanceCondition: core.condition.Condition,
+    instance: core.instance.Instance,
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/raml/PodStatusConversion.scala]: What is the motivation for this change?

File: src/test/scala/mesosphere/marathon/MarathonSchedulerActorTest.scala [hunk 6]
```diff
@@ -140,7 +140,6 @@ class MarathonSchedulerActorTest extends AkkaUnitTest with ImplicitSender with G
         .addTaskDropped(containerName = Some("dropped"))
         .addTaskUnknown(containerName = Some("unknown"))
         .addTaskKilling(containerName = Some("killing"))
-        .addTaskReserved(containerName = Some("reserved"))
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/MarathonSchedulerActorTest.scala]: We should keep this test. There are still reserved tasks! We dropped the condition not task with reservations.

File: src/test/scala/mesosphere/marathon/core/instance/TestInstanceBuilder.scala [hunk 7]
```diff
@@ -31,12 +31,6 @@ case class TestInstanceBuilder(instance: Instance, now: Timestamp = Timestamp.no
   def addTaskLost(since: Timestamp = now, containerName: Option[String] = None): TestInstanceBuilder =
     addTaskWithBuilder().taskLost(since, containerName).build()
 
-  def addTaskReserved(containerName: Option[String] = None): TestInstanceBuilder =
-    addTaskWithBuilder().taskReserved(containerName).build()
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/core/instance/TestInstanceBuilder.scala]: As I mentioned above we still have to have these. We did not remove reserved tasks.

File: src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdater.scala (hunk 1 above)
Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdater.scala]: hmm I don't like the name `Reserved` to be honest, because in general sense reserved does not mean terminal. Instance that wasn't yet launched is also reserved in a sense...

But I would be totally open to other suggestions

File: src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueActor.scala (hunk 2 above)
Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueActor.scala]: hmm that's a fair point! because now we would reschedule tasks even before they were killed.

File: src/main/scala/mesosphere/marathon/core/task/update/impl/steps/ScaleAppUpdateStepImpl.scala (hunk 4 above)
Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/core/task/update/impl/steps/ScaleAppUpdateStepImpl.scala]: after phase 2 reservedterminal and reserved was the same see https://github.com/mesosphere/marathon/pull/6658/files#diff-4c766f1e11cd8f4e0f48c1a4392b2b41L43

File: src/main/scala/mesosphere/marathon/raml/PodStatusConversion.scala (hunk 5 above)
Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/raml/PodStatusConversion.scala]: I needed to do that because of this line https://github.com/mesosphere/marathon/pull/6658/files#diff-e0df4221bad191a89f0d04c9d30bbd06R253

File: src/test/scala/mesosphere/marathon/MarathonSchedulerActorTest.scala (hunk 6 above)
Comment (alenkacz) [reply src/test/scala/mesosphere/marathon/MarathonSchedulerActorTest.scala]: well this piece of code really works with conditions only so it does not make sense to add residents here IMHO. But in general yes, this PR is definitely missing unit tests coverage. I'll follow up with that once I debug the integration test failures. Feel free to come up with something when I am sleeping :-) I am more than willing to collaborate on this

File: src/main/scala/mesosphere/marathon/raml/PodStatusConversion.scala (hunk 5 above)
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/raml/PodStatusConversion.scala]: Got it! Thanks!

File: docs/docs/rest-api/public/api/v2/schema/AppDefinition.json [hunk 8]
```diff
@@ -120,7 +120,7 @@
           },
           {
             "type": "string",
-            "enum": ["UNIQUE", "CLUSTER", "GROUP_BY", "LIKE", "UNLIKE", "MAX_PER"]
+            "enum": ["UNIQUE", "CLUSTER", "GROUP_BY", "LIKE", "UNLIKE", "MAX_PER", "IS"]
```
Comment (jeschkies) [inline docs/docs/rest-api/public/api/v2/schema/AppDefinition.json]: Could you rebase or merge on master. I don't think this should be part of your diff.

File: src/test/scala/mesosphere/marathon/core/instance/TestInstanceBuilder.scala (hunk 7 above)
Comment (alenkacz) [reply src/test/scala/mesosphere/marathon/core/instance/TestInstanceBuilder.scala]: these builders just created instances and tasks with condition reserved, this is really obsolete and we don't need that anymore since we dropped that condition.

File: src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueActor.scala [hunk 9]
```diff
@@ -270,7 +270,7 @@ private[impl] class LaunchQueueActor(
 
       // Reuse resident instances that are stopped.
       val existingReservedStoppedInstances = await(instanceTracker.specInstances(runSpec.id))
-        .filter(i => i.isReserved && i.state.goal == Goal.Stopped) // resident to relaunch
+        .filter(i => i.hasReservation && i.state.condition.isTerminal && i.state.goal == Goal.Stopped) // resident to relaunch
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueActor.scala]: Hm, I think we can drop the `i.state.condition.isTerminal` in the future.

File: src/main/scala/mesosphere/marathon/raml/PodStatusConversion.scala [hunk 10]
```diff
@@ -249,36 +247,39 @@ trait PodStatusConversion {
...
-      case condition.Condition.Running =>
-        if (containerStatus.exists(_.conditions.exists { cond =>
-          cond.name == STATUS_CONDITION_HEALTHY && cond.value == "false"
-        }))
-          PodInstanceState.Degraded -> Some(MSG_INSTANCE_UNHEALTHY_CONTAINERS)
-        else
-          PodInstanceState.Stable -> None
+    if (instance.isScheduled) {
+      PodInstanceState.Pending -> None
+    } else {
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/raml/PodStatusConversion.scala]: I don't quite understand this change. We are checking for the scheduled condition below so why the extra if clause?

File: src/test/scala/mesosphere/marathon/MarathonSchedulerActorTest.scala (hunk 6 above)
Comment (jeschkies) [reply src/test/scala/mesosphere/marathon/MarathonSchedulerActorTest.scala]: I don't know. I mean we check `assert(nonTerminalTasks.size == 5, "We should have 5 non-terminal tasks")` below. With this in still in place it should be the same, right? In any case I don't understand why you dropped this case.

File: src/test/scala/mesosphere/marathon/core/instance/TestInstanceBuilder.scala (hunk 7 above)
Comment (jeschkies) [reply src/test/scala/mesosphere/marathon/core/instance/TestInstanceBuilder.scala]: Hm, how is a `TestTaskBuilder.taskResidentLaunched` different from `TestTaskBuilder.taskLaunched`? Seems we should `TestTaskBuilder.taskResidentLaunched` as well then.

File: src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdater.scala (hunk 1 above)
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdater.scala]: Ok, let's leave it for now.

File: src/main/scala/mesosphere/marathon/raml/PodStatusConversion.scala (hunk 10 above)
Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/raml/PodStatusConversion.scala]: look at how was `Reserved` used in the previous diff. For reserved, we showed state as pending. `Reserved` condition is dropped now. When thinking in general, I believe that`Pending` is the correct `PodInstanceState` for instances that are SCHEDULED - meaning `isScheduled` is true

File: src/test/scala/mesosphere/marathon/MarathonSchedulerActorTest.scala (hunk 6 above)
Comment (alenkacz) [reply src/test/scala/mesosphere/marathon/MarathonSchedulerActorTest.scala]: I tend to disagree. This code path is really working only with conditions and verifying condition mapping. You might argue that we should also test task with a long name, task with some label or task with whatever (e.g. reservation). Task does not even have reservation, that's instance. I really believe it does not make sense keeping it here.

File: src/test/scala/mesosphere/marathon/core/instance/TestInstanceBuilder.scala (hunk 7 above)
Comment (alenkacz) [reply src/test/scala/mesosphere/marathon/core/instance/TestInstanceBuilder.scala]: `taskResidentLaunched` creates *instance with reservation* and a task that is running. I actually think that we are mixing two things there because reservation is really not a property of task. But I don't think it makes sense to invest more into those builders other than removing them and replacing them with production code

Comment (jeschkies) [reply src/test/scala/mesosphere/marathon/core/instance/TestInstanceBuilder.scala]: Hm, why not fix it? It's not a blocker for this PR but should it be 

```scala
TestInstanceBuilder(...).withReservation(...).addRunningTask()
```
then?

File: src/main/scala/mesosphere/marathon/raml/PodStatusConversion.scala (hunk 10 above)
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/raml/PodStatusConversion.scala]: Ah, that relates to #6616 again where `condition = Scheduled` is not enough but one has to check `isScheduled`. Got it. Thanks for clarifying.

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo18100.scala [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+  import mesosphere.marathon.api.v2.json.Formats.TimestampFormat
+
+  /**
+    * Read format for instance state with reserved condition.
+    */
+  val migrationConditionReader = new Reads[Condition] {
+    private def readString(j: JsReadable) = j.validate[String].map {
+
+      case reserved if reserved.toLowerCase == "reserved" => Condition.Finished
+      case reservedTerminal if reservedTerminal.toLowerCase == "reservedterminal" => Condition.Finished
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/storage/migration/MigrationTo18100.scala]: Infer condition from Mesos status.

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo17.scala [hunk 12]
```diff
@@ -134,7 +134,7 @@ object MigrationTo17 extends MaybeStore with StrictLogging {
     */
   def migrateInstanceGoals(instanceRepository: InstanceRepository, persistenceStore: PersistenceStore[_, _, _])(implicit mat: Materializer): Future[Done] = {
 
-    logger.info("Starting goal migration to Storage version 200")
+    logger.info("Starting goal migration to Storage version 17")
```
Comment (timcharper) [inline src/main/scala/mesosphere/marathon/storage/migration/MigrationTo17.scala]: Was this chance intentional?

Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/storage/migration/MigrationTo17.scala]: Yes. The 200 was wrong.
